packages/maskbook/src/plugins/Voicechat/UI/VoicechatInlay.tsx

```diff
@@ -0,0 +1,200 @@
+    const { t } = useI18N()
+
+    let elements: PeerMediaElement[] = []
+
+    const [speakingUsernames, setSpeakingUsernames] = useState<string[]>([])
+
+    registerSpeakingUsernames((speakingUsernames: string[]) => setSpeakingUsernames(speakingUsernames))
+
+    if (props.peerMediaElements) {
+        const peerMediaElements = props.peerMediaElements
+        elements = Object.keys(peerMediaElements).map((peerMediaKey) => peerMediaElements[peerMediaKey])
+    }
+    if (props.localPeerMediaElement) {
+        elements.push(props.localPeerMediaElement)
+    }
+
+    return (
+        <div>
+            {elements.map((peerMedia) => {
+                const isSpeaking = speakingUsernames.includes(peerMedia.username)
+                const SpeakIcon = () => (isSpeaking ? <VolumeUpOutlined /> : <VolumeMuteOutlined />)
+
+                return (
+                    <Box key={peerMedia.username + peerMedia.isLocal} display={'flex'} alignItems={'center'}>
+                        <Box mr={0.5} style={{ marginTop: '6px' }}>
+                            <SpeakIcon />
+                        </Box>
+                        <Typography component="p" sx={{ fontWeight: isSpeaking ? 'bold' : 'regular' }}>
+                            {peerMedia.isLocal ? t('plugin_voicechat_you') : peerMedia.username}
+                        </Typography>
+                    </Box>
+                )
+            })}
+        </div>
+    )
+}
+
+export function VoicechatInlay(props: VoicechatInlayProps) {
+    const postIdentifier = usePostInfoDetails('postIdentifier')
+    const postBy = usePostInfoDetails('postBy')
+
+    const [peerMediaElements, setPeerMediaElements] = useState<Record<string, PeerMediaElement>>()
+    const [localPeerMediaElement, setLocalPeerMediaElement] = useState<PeerMediaElement | null>()
+    const [mediaDevices, setMediaDevices] = useState<MediaDeviceInfo[]>([])
+    const [activeMediaDevice, setActiveMediaDevice] = useState<string>('')
+    const [isMuted, setIsMuted] = useState<boolean>(false)
+    const [isJoining, setIsJoining] = useState<boolean>(false)
+
+    const { t } = useI18N()
+
+    if (!postIdentifier) {
+        return null
+    }
+
+    const joinChat = () => {
+        setIsJoining(true)
+
+        navigator.mediaDevices.enumerateDevices().then((devices) => {
+            setMediaDevices([...devices.filter((d) => d.kind === 'audioinput')])
+        })
+
+        const customServer = props.metadata.customServer ? props.metadata.customServer : ''
+
+        init(
+            customServer,
+            postIdentifier.postId,
+            postBy.userId,
+            (peerMedia, localPeer) => {
+                setPeerMediaElements({ ...peerMedia })
+                setLocalPeerMediaElement(localPeer)
+                setIsJoining(false)
+            },
+            (activeMediaDevice: string) => {
+                setActiveMediaDevice(activeMediaDevice)
+            },
+        )
+    }
+
+    const JoinButton = () => {
```

Comment:
Hi please do not define nested components. That will cause significant performance problems and real bugs. You may want to learn more details on https://reactjs.org/docs/reconciliation.html#the-diffing-algorithm